vms/proposervm/block/build.go

```diff
@@ -6,39 +6,152 @@ package block
+	// signature, which is prefixed by a uint32. So, we need to strip off the
+	// signature as well as its length prefix to get the unsigned bytes.
+	lenUnsignedBytes := len(bytes) - wrappers.IntLen - len(signature)
+
+	if lenUnsignedBytes < 0 {
+		return ids.Empty, errInvalidBlockEncodingLength
+	}
+
+	unsignedBytes = bytes[:lenUnsignedBytes]
+	return hashing.ComputeHash256Array(unsignedBytes), nil
+}
+
+// marshalBlock marshal the given statelessBlock by using either the default statelessBlock or
+// coping the exported fields into statelessBlockV0 and then marshaling it.
+// this allows the marsheler to produce encoded blocks that match the old style blocks as long as
+// the VRFSig feature was not enabled.
+func marshalBlock(block *statelessBlock[statelessUnsignedBlock]) ([]byte, error) {
+	if len(block.StatelessBlock.VRFSig) == 0 {
+		// create a backward compatible block ( without VRFSig ) and use the statelessBlockV0 encoder for the encoding.
+		var preBlockSigBlock SignedBlock = &statelessBlock[statelessUnsignedBlockV0]{
+			StatelessBlock: statelessUnsignedBlockV0{
+				ParentID:     block.StatelessBlock.ParentID,
+				Timestamp:    block.StatelessBlock.Timestamp,
+				PChainHeight: block.StatelessBlock.PChainHeight,
+				Certificate:  block.StatelessBlock.Certificate,
+				Block:        block.StatelessBlock.Block,
+			},
+			Signature: block.Signature,
+		}
+		return Codec.Marshal(CodecVersion, &preBlockSigBlock)
+	}
+	var blockIntf SignedBlock = block
+	return Codec.Marshal(CodecVersion, &blockIntf)
+}
+
+func calculateBootstrappingBlockSig(chainID ids.ID, networkID uint32) [hashing.HashLen]byte {
+	// build the hash of the following struct:
+	// +-----------------------+----------+------------+
+	// |  prefix :             | [8]byte  | "rng-root" |
+	// +-----------------------+----------+------------+
+	// |  chainID :            | [32]byte |  32 bytes  |
+	// +-----------------------+----------+------------+
+	// |  networkID:           | uint32   |  4 bytes   |
+	// +-----------------------+----------+------------+
+
+	buffer := make([]byte, len(vrfRngRootPrefix)+ids.IDLen+4)
+	copy(buffer, vrfRngRootPrefix)
+	copy(buffer[len(vrfRngRootPrefix):], chainID[:])
+	binary.LittleEndian.PutUint32(buffer[len(vrfRngRootPrefix)+ids.IDLen:], networkID)
+	return hashing.ComputeHash256Array(buffer)
+}
+
+func NextHashBlockSignature(parentBlockSig []byte) []byte {
+	if len(parentBlockSig) == 0 {
+		return nil
+	}
+	// previous block had a valid signature, hash that signature.
+	return hashing.ComputeHash256(parentBlockSig)
+}
+
+func NextBlockVRFSig(parentBlockVRFSig []byte, blsSignKey *bls.SecretKey, chainID ids.ID, networkID uint32) []byte {
+	if blsSignKey == nil {
+		// if we need to build a block without having a BLS key, we'll be hashing the previous
+		// signature only if it presents. Otherwise, we'll keep it empty.
+		if len(parentBlockVRFSig) == 0 {
+			// no parent block signature.
+			return []byte{}
```

Comment:
Nit: do we need to return `[]byte{}` instead of `nil`? Typically I think we return the latter, unless `[]byte{}` has a special meaning distinct from `nil`.

Reply:
done